### PR TITLE
Callback

### PR DESCRIPTION
**Pull request by ibmmqmet**
Please ensure all items are complete before opening.

- [x] Tick to sign-off your agreement to the [IBM Contributor License Agreement](https://github.com/ibm-messaging/mq-golang/CLA.md)
- [x] You have added tests for any code changes
- [x] You have updated the [CHANGES.md](https://github.com/ibm-messaging/mq-golang/CHANGES.md)
- [x] You have completed the PR template below:

## What

Added MQCB/MQCTL implementation
Added MQBEGIN to complete the set
Modified sample programs for more appropriate coding.

## How

Follow usual pattern for new structures. Wrap the user-supplied callback function with a Go function which in turn is wrapped by a C function. Go callbacks from C are not very well documented, but this works.

Sample programs updated to use "defer" instead of simply suggesting them, to encourage better design. FAIL_IF_QUIESCING no longer needed in user apps as it will get set automatically.

## Testing
New sample amqscb demonstrates the callback mechanism
How to test your changes work, not required for documentation changes.

## Issues

Links to the github issue(s) (if present) that this pull request is resolving.

### CODE REVIEW

File: ibmmq/mqiCBD.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	return cbd
+}
+
+func copyCBDtoC(mqcbd *C.MQCBD, gocbd *MQCBD) {
+
+	setMQIString((*C.char)(&mqcbd.StrucId[0]), "CBD ", 4)
+	mqcbd.Version = C.MQCBD_VERSION_1
+
+	mqcbd.CallbackType = C.MQLONG(gocbd.CallbackType)
+	mqcbd.Options = C.MQLONG(gocbd.Options) | C.MQCBDO_FAIL_IF_QUIESCING
```
Comment (parrobe) [inline ibmmq/mqiCBD.go]: Same here. Always setting `MQCBDO_FAIL_IF_QUIESCING`

File: ibmmq/mqicb.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+		return &mqreturn
+	}
+
+	return nil
+}
+
+// Functions below here manage the map of objects and control information so that
+// the Go variables can be saved/restored from invocations to the C layer
+func makeKey(hConn C.MQHCONN, hObj C.MQHOBJ) string {
+	key := fmt.Sprintf("%d/%d", hConn, hObj)
```
Comment (parrobe) [inline ibmmq/mqicb.go]: Could just do: `return fmt.Sprintf("%d/%d", hConn, hObj)` instead of setting variable to return it.

File: ibmmq/mqicb.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	return key
+}
+
+func makePartialKey(hConn C.MQHCONN) string {
+	key := fmt.Sprintf("%d/", hConn)
```
Comment (parrobe) [inline ibmmq/mqicb.go]: Same as above.

File: ibmmq/mqi.go [hunk 4]
```diff
@@ -248,7 +252,7 @@ func (x *MQQueueManager) Open(good *MQOD, goOpenOptions int32) (MQObject, error)
 	}
 
 	copyODtoC(&mqod, good)
-	mqOpenOptions = C.MQLONG(goOpenOptions)
+	mqOpenOptions = C.MQLONG(goOpenOptions) | C.MQOO_FAIL_IF_QUIESCING
```
[the remark this answers is not in the document]
Comment (ibmmqmet) [reply ibmmq/mqi.go]: FAIL_IF_QUIESCING should always be set - it's one of the big complaints about it not being a default. There's no good reason not to use it in applications. So forcing it here makes user applications - and the sample programs - simpler as they don't need to think about it (assuming they ever did, and then wondered why qmgrs didn't end cleanly or restart). I did similar in the Node.js implementation based on user feedback. 

Yes, the code could return "key" directly but setting a variable made it easier to insert Printfs during debug.

**Pull request merged**